[docs] Add docs for adding a new backend (#2878)
NVIDIA/cuda-quantum: docs/sphinx/using/backends/new_backends.rst

Reviewer: I think this page should be under the 'Extending' section of the docs. We can add a line from the [Backends](https://nvidia.github.io/cuda-quantum/latest/using/backends/backends.html) page pointing to it (to add a new hardware backend please see...)

*************************
Adding New Backends
*************************

This guide explains how to integrate a new quantum hardware provider with CUDA-Q. The integration process involves creating a server helper that handles communication with the provider's API, defining configuration files, and implementing necessary tests.

Overview
========

CUDA-Q supports various quantum hardware providers through a REST-based architecture. To add a new backend, you'll need to:

1. Create a server helper class that handles API communication
2. Define configuration files for the target
3. Implement tests to verify functionality
4. Add documentation for users

The following sections detail each step of this process.

Server Helper Implementation
===========================

> The server helper is the core component that handles communication with the quantum hardware provider's API. It extends the ``ServerHelper`` base class and implements methods for job submission, result retrieval, and other provider-specific functionality.

Reviewer: add link to the code definition of Server Helper

        std::string path = "/jobs";
        
        return std::make_tuple(backendConfig["url"] + path, headers, 
                              std::vector<ServerMessage>{job});
      }
    
      /// @brief Example implementation of job ID tracking.
      std::string extractJobId(ServerMessage &postResponse) override {
        if (!postResponse.contains("id"))
          return "";
        
        return postResponse.at("id");
      }
    
      /// @brief Example implementation of job ID tracking.
      std::string constructGetJobPath(ServerMessage &postResponse) override {
        return extractJobId(postResponse);
      }
    
      /// @brief Example implementation of job ID tracking.
      std::string constructGetJobPath(std::string &jobId) override {
        return backendConfig["url"] + "/jobs/" + jobId;
      }
    
      /// @brief Example implementation of job status checking.
      bool jobIsDone(ServerMessage &getJobResponse) override {
        if (!getJobResponse.contains("status"))
          return false;
        
        std::string status = getJobResponse["status"];
        return status == "COMPLETED" || status == "FAILED";
      }
    

>       /// @brief Example implementation of result processing.

Reviewer 2: ```suggestion
      /// @brief Example implementation of result processing. 
      The raw results from quantum hardware often need post-processing (bit reordering, normalization, etc.) to match CUDA-Q's expectations. Handle it here.
```

        cudaq::info("Processing results: {}", getJobResponse.dump());
        
        // Extract measurement results from the response
        auto samplesJson = getJobResponse["results"]["counts"];
        cudaq::CountsDictionary counts;
        
        for (auto &item : samplesJson.items()) {
          std::string bitstring = item.key();
          std::size_t count = item.value();
          counts[bitstring] = count;
        }
        
        // Create an ExecutionResult
        cudaq::ExecutionResult execResult{counts};
        
        // Return the sample_result
        return cudaq::sample_result{execResult};
      }
    
      /// @brief Example implementation of polling configuration.
      std::chrono::microseconds
      nextResultPollingInterval(ServerMessage &postResponse) override {
        return std::chrono::seconds(5);
      }
    };
    
    } // namespace cudaq
    
    // Register the server helper in the CUDA-Q server helper factory
    CUDAQ_REGISTER_TYPE(cudaq::ServerHelper, cudaq::ProviderNameServerHelper, <provider_name>)

``CMakeLists.txt``
--------------

Create a ``CMakeLists.txt`` file for your server helper:

> 

Reviewer 2: With newly added targets we want to make them opt-in with default behavior being that they are always built in CUDA-Q. This will help developers build specific targets and skip them if not working on those aspects.
Suggest to add following -
```
To make your target optional, add a CMake flag in the top-level ``CMakeLists.txt``:

.. code-block:: cmake

    # Enable <provider_name> target by default
    if (NOT DEFINED CUDAQ_ENABLE_PROVIDER_NAME_BACKEND)
      set(CUDAQ_ENABLE_PROVIDER_NAME_BACKEND ON CACHE BOOL "Enable building the <Provider Name> target.")
    endif()

Then, in your target's ``CMakeLists.txt``, check for this flag:

.. code-block:: cmake

    if(CUDAQ_ENABLE_PROVIDER_NAME_BACKEND)
      target_sources(cudaq-rest-qpu PRIVATE ProviderNameServerHelper.cpp)
      add_target_config(<provider_name>)
      
      add_library(cudaq-serverhelper-<provider_name> SHARED ProviderNameServerHelper.cpp)
      target_link_libraries(cudaq-serverhelper-<provider_name>
        PUBLIC
        cudaq-common
        fmt::fmt-header-only
      )
      install(TARGETS cudaq-serverhelper-<provider_name> DESTINATION lib)
    endif()
```

Python tests can be skipped if the target is not built -
```
skipIf<provider>NotInstalled = pytest.mark.skipif(
    not (cudaq.has_target("<provider>")),
    reason='Could not find `<provider>` in installation')
```

C++ test can be checked with following addition to `` targettests/lit.site.cfg.py.in`` file
```
config.cudaq_backends_provider = "@CUDAQ_ENABLE_PROVIDER_NAME_BACKEND@"
if cmake_boolvar_to_bool(config.cudaq_backends_provider):
  config.available_features.add('provider')
  config.substitutions.append(('%provider_avail', 'true'))
else:
  config.substitutions.append(('%provider_avail', 'false'))
```
and
`// RUN: if %provider_avail; then nvq++ %cpp_std --target provider %s -o %t.x; fi`

.. code-block:: cmake

    # CMakeLists.txt
    target_sources(cudaq-rest-qpu PRIVATE ProviderNameServerHelper.cpp)
    add_target_config(<provider_name>)
    
    add_library(cudaq-serverhelper-<provider_name> SHARED ProviderNameServerHelper.cpp)
    target_link_libraries(cudaq-serverhelper-<provider_name>
      PUBLIC
      cudaq-common
      fmt::fmt-header-only
    )
    install(TARGETS cudaq-serverhelper-<provider_name> DESTINATION lib)

Target Configuration
===================

Create a ``YAML`` configuration file for your target:

.. code-block:: yaml

    # <provider_name>.yml
    name: "<provider_name>"
    description: "CUDA-Q target for Provider Name."
    
    config:
      # Tell DefaultQuantumPlatform what QPU subtype to use
      platform-qpu: remote_rest
      # Add the rest-qpu library to the link list
      link-libs: ["-lcudaq-rest-qpu"]
      # Tell NVQ++ to generate glue code to set the target backend name
      gen-target-backend: true
      # Add preprocessor defines to compilation
      preprocessor-defines: ["-D CUDAQ_QUANTUM_DEVICE"]
      # Define the lowering pipeline

>       # You will likely need assistance from the CUDA-Q team to set this up correctly

Reviewer 2: +1

```suggestion
      # This will cover applying hardware-specific constraints since each provider may have different native gate sets, requiring custom mappings and decompositions. You may need assistance from the CUDA-Q team to set this up correctly.
```

    TEST_STATUS=$?
    
    # Kill the server
    kill $SERVER_PID
    
    # Return the test status
    exit $TEST_STATUS

4. Implement the C++ test:

.. code-block:: cpp

    // ProviderNameTester.cpp
    #include "common/Logger.h"
    #include "common/RestClient.h"
    #include "common/ServerHelper.h"
    #include "cudaq/platform/quantum_platform.h"
    #include "gtest/gtest.h"
    
    TEST(ProviderNameTester, checkSimpleCircuit) {
      // Initialize the platform
      auto platform = cudaq::get_platform();
      platform->setTargetBackend("<provider_name>");
      
      // Set configuration
      platform->setBackendParameter("url", "http://localhost:PORT");
      platform->setBackendParameter("api_key", "test_key");
      
      // Create a simple circuit
      auto kernel = cudaq::make_kernel();
      auto qubits = kernel.qalloc(2);
      kernel.h(qubits[0]);
      kernel.cx(qubits[0], qubits[1]);
      kernel.mz(qubits);
      
      // Execute the circuit
      auto counts = cudaq::sample(kernel);
      
      // Check results
      EXPECT_EQ(counts.size(), 2);
      EXPECT_TRUE(counts.has_key("00"));
      EXPECT_TRUE(counts.has_key("11"));
    }

> Mock Server

Reviewer: You might want to explain what the mock server is and how it is being used.

            else:
                self._set_headers(404)
                self.wfile.write(json.dumps({'error': 'Not found'}).encode())
    
        def do_GET(self):
            if self.path.startswith('/jobs/job-123'):
                # Return job status and results
                response = {
                    'id': 'job-123',
                    'status': 'COMPLETED',
                    'results': {
                        'counts': {
                            '00': 500,
                            '11': 500
                        }
                    }
                }
                self._set_headers()
                self.wfile.write(json.dumps(response).encode())
            else:
                self._set_headers(404)
                self.wfile.write(json.dumps({'error': 'Not found'}).encode())
    
    def startServer(port=8000):
        server_address = ('', port)
        httpd = HTTPServer(server_address, ProviderNameMockServer)
        print(f'Starting mock server on port {port}...')
        httpd.serve_forever()
    
    if __name__ == '__main__':
        port = int(sys.argv[1]) if len(sys.argv) > 1 else 8000

>         startServer(port)

Author: We can... but I'd rather not scope creep on documenting the current process vs overhauling the process.

Python Tests
-----------

Create Python tests for your backend:

.. code-block:: python

    # python/tests/backends/test_<provider_name>.py
    import os
    import sys
    import time
    import pytest
    from multiprocessing import Process
    
    import cudaq
    from cudaq import spin
    
    try:
        from utils.mock_qpu.<provider_name> import startServer
    except:
        print("Mock qpu not available, skipping Provider Name tests.")
        pytest.skip("Mock qpu not available.", allow_module_level=True)
    
    # Define the port for the mock server

>     port = 62444

Reviewer: Is it always the same port or does each QPU needs their own port?

    
    @pytest.fixture(scope="session", autouse=True)
    def startUpMockServer():
        # Set the targeted QPU
        cudaq.set_target('provider_name',
                        url=f'http://localhost:{port}',
                        api_key="test_key")
        
        # Launch the Mock Server
        p = Process(target=startServer, args=(port,))
        p.start()
        time.sleep(1)
        
        yield "Running the tests."
        
        # Kill the server
        p.terminate()
    
    def test_provider_name_sample():
        # Create the kernel
        kernel = cudaq.make_kernel()
        qubits = kernel.qalloc(2)
        kernel.h(qubits[0])
        kernel.cx(qubits[0], qubits[1])
        kernel.mz(qubits)
        
        # Run sample
        counts = cudaq.sample(kernel)
        assert len(counts) == 2
        assert '00' in counts
        assert '11' in counts
        
        # Run sample asynchronously
        future = cudaq.sample_async(kernel)
        counts = future.get()
        assert len(counts) == 2
        assert '00' in counts

>         assert '11' in counts

Author: Ditto. 😅

> Documentation

Reviewer 2: Should we add a section to specifically call out integration testing requirements?
```
To ensure proper execution on the hardware, a validation backend must be provided that:

1. Consumes the same format that your target will use
2. Validates that circuits passing this validation will execute successfully on the actual hardware
3. Can be accessed by CUDA-Q's GitHub CI/CD pipelines

Your validation backend doesn't need to be publicly available, but it should:

- Accept the same input format as your actual quantum processor
- Return meaningful error messages for invalid circuits
- Provide an API endpoint that can be called from our integration tests
```

followed by how-to
```
Add your target to ``.github/workflows/integration_tests.yml``:

.. code-block:: yaml

    - name: Submit to <provider_name> test server
      if: (success() || failure()) && (inputs.target == '<provider_name>' || github.event_name == 'schedule')
      run: |
        echo "### Submit to <provider_name> server" >> $GITHUB_STEP_SUMMARY
        # Set up any required dependencies        
        # Set up environment variables for authentication
        export PROVIDER_API_KEY='${{ secrets.PROVIDER_API_KEY }}'
        
        # Run the integration tests
        python_tests="docs/sphinx/targets/python/<provider_name>.py"
        cpp_tests="docs/sphinx/targets/cpp/<provider_name>.cpp"
        
        # Execute tests (see other provider examples for implementation details)
        # ...
```

============

Add documentation for your backend in the appropriate sections of the CUDA-Q documentation. This should include:

> 

Reviewer 2: ```suggestion
0. How to access your server (authentication set up, documentation etc.)
```
?

1. How to configure and use the backend
2. Any provider-specific parameters or features
3. Examples of running circuits on the backend

> 

Reviewer: @mawolf2023 any requirements on adding the icon to the backends [diagram](https://nvidia.github.io/cuda-quantum/latest/using/backends/hardware.html) / list ? (Alphabetical order per modality, specific icon format, etc. )

Reviewer 3: @efratshabtai  I think that is it.  Just making sure to have an accurate logo and then put them alphabetical under each modality.  I can share the file to make the edits too.

More specifically, you will need to modify at least the following files:
* ``docs/sphinx/using/examples/hardware_providers.rst``
* ``docs/sphinx/using/backends/hardware/<your-technology>.rst``
* ``docs/sphinx/targets/python/<provider_name>.py``
* ``docs/sphinx/targets/cpp/<provider_name>.cpp``

Example Usage
============

Once your backend is implemented, users can use it as follows:

.. code-block:: python

    import cudaq
    
    # Set the target to your provider
    cudaq.set_target('<provider_name>', 
                    api_key='your_api_key',
                    device='your_device')
    
    # Create and run a circuit
    @cudaq.kernel
    def bell():
        qubits = cudaq.qvector(2)
        h(qubits[0])
        x.ctrl(qubits[0], qubits[1])
        mz(qubits)
    
    # Run the circuit
    counts = cudaq.sample(bell)
    print(counts)

> Conclusion

Reviewer: Let's add a section of maintenance: we want the integration to continue to work with new CUDA-Q versions and new hardware systems - what is under the CUDA-Q team responsibility? what is under the responsibility of the QPU provider?

==========

> By following this guide, you can integrate a new quantum hardware provider with CUDA-Q. The integration involves creating a server helper, defining configuration files, and implementing tests. Once integrated, users can seamlessly run quantum circuits on your provider's hardware using the CUDA-Q framework.

Reviewer: Any instructions on how to create a PR with these changes?
